### ISSUE TITLE
[ibm-telco] haproxy service racing over httpd service

### ISSUE THREAD

**Issue by maheshd2** [redhat-ztp/ztp-cluster-deploy#87]
We were trying to deploy cluster using stable/4.7 branch. And found that httpd service was failing to start.

Sequence of steps to reproduce:

1. Ran 'prepare-environment' role.
2. Manually started httpd service "systemctl start httpd " (Workaround for another issue https://github.com/redhat-ztp/ztp-cluster-deploy/issues/48)
3. Ran create-cluster and enroll-hosts roles.
4. Rebooted Master Host Server.
5. Ran deploy-cluster, expose-endpoints, create-cluster-day2 and modify-iso-day2 roles.
6. Ran add-remote-worker role - This role failed because httpd service was not running to serve the ISO image.

Found httpd server was failing to start:
```
[root@vcuhost ai-deploy-cluster-remoteworker]# service httpd status
Redirecting to /bin/systemctl status httpd.service
● httpd.service - The Apache HTTP Server
   Loaded: loaded (/usr/lib/systemd/system/httpd.service; disabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Thu 2021-03-25 22:35:39 EDT; 20s ago
     Docs: man:httpd.service(8)
  Process: 29140 ExecStart=/usr/sbin/httpd $OPTIONS -DFOREGROUND (code=exited, status=1/FAILURE)
 Main PID: 29140 (code=exited, status=1/FAILURE)
   Status: "Reading configuration..."

Mar 25 22:35:39 vcuhost.rtp.raleigh.ibm.com systemd[1]: Starting The Apache HTTP Server...
Mar 25 22:35:39 vcuhost.rtp.raleigh.ibm.com httpd[29140]: (98)Address already in use: AH00072: make_sock: could not bind to address [::>
Mar 25 22:35:39 vcuhost.rtp.raleigh.ibm.com httpd[29140]: (98)Address already in use: AH00072: make_sock: could not bind to address 0.0>
Mar 25 22:35:39 vcuhost.rtp.raleigh.ibm.com httpd[29140]: no listening sockets available, shutting down
Mar 25 22:35:39 vcuhost.rtp.raleigh.ibm.com httpd[29140]: AH00015: Unable to open logs
Mar 25 22:35:39 vcuhost.rtp.raleigh.ibm.com systemd[1]: httpd.service: Main process exited, code=exited, status=1/FAILURE
Mar 25 22:35:39 vcuhost.rtp.raleigh.ibm.com systemd[1]: httpd.service: Failed with result 'exit-code'.
Mar 25 22:35:39 vcuhost.rtp.raleigh.ibm.com systemd[1]: Failed to start The Apache HTTP Server.
``` 

Then I found that haproxy was running on port 80.
```
[root@vduhost ai-deploy-cluster-remoteworker]# netstat -anlp | grep 80
tcp        0      0 0.0.0.0:8000            0.0.0.0:*               LISTEN      5225/conmon
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      28029/haproxy
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      5225/conmon
tcp        0      0 0.0.0.0:8090            0.0.0.0:*               LISTEN      5225/conmon
tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN      28029/haproxy
udp        0      0 0.0.0.0:50642           0.0.0.0:*                           28027/haproxy-maste
unix  2      [ ACC ]     SEQPACKET  LISTENING     19591    5225/conmon          /var/run/libpod/socket/f5ceb14fca87a6a0102f47ff2d76d49d334efcce6f12fe3980f8739ec26f1b8d/attach
unix  2      [ ACC ]     SEQPACKET  LISTENING     21306    27996/conmon         /var/run/libpod/socket/886e4366a409a32fade94b09c26dd2b239604b0052e44b68449858058841259a/attach
unix  3      [ ]         STREAM     CONNECTED     67804    3568/dbus-daemon     /run/user/42/bus
unix  3      [ ]         STREAM     CONNECTED     78016    1/systemd            /run/systemd/journal/stdout
unix  3      [ ]         STREAM     CONNECTED     92680    4067/gsd-screensave
unix  3      [ ]         STREAM     CONNECTED     78006    1/systemd            /run/systemd/journal/stdout
unix  3      [ ]         STREAM     CONNECTED     55804    4058/gsd-keyboard
unix  3      [ ]         STREAM     CONNECTED     78004    1/systemd            /run/systemd/journal/stdout
unix  2      [ ]         DGRAM                    46206    3080/rsyslogd
unix  3      [ ]         STREAM     CONNECTED     55802    4058/gsd-keyboard
unix  3      [ ]         STREAM     CONNECTED     86803    1/systemd            /run/systemd/journal/stdout
unix  3      [ ]         STREAM     CONNECTED     78002    1/systemd            /run/systemd/journal/stdout
unix  3      [ ]         STREAM     CONNECTED     55800    4058/gsd-keyboard
unix  2      [ ]         STREAM     CONNECTED     17180    2448/polkitd
unix  3      [ ]         STREAM     CONNECTED     67802    3993/ibus-portal
unix  2      [ ]         DGRAM                    61804    3583/virtlogd
unix  3      [ ]         STREAM     CONNECTED     52380    3285/Xorg            @/tmp/.X11-unix/X0
unix  3      [ ]         STREAM     CONNECTED     67806    3568/dbus-daemon     /run/user/42/bus
unix  3      [ ]         STREAM     CONNECTED     30980    4083/gsd-wacom
unix  3      [ ]         STREAM     CONNECTED     71803    2460/dbus-daemon     /run/dbus/system_bus_socket
unix  3      [ ]         STREAM     CONNECTED     77980    1/systemd            /run/systemd/journal/stdout
unix  3      [ ]         STREAM     CONNECTED     67803    3993/ibus-portal
unix  3      [ ]         STREAM     CONNECTED     58580    4012/upowerd
unix  3      [ ]         STREAM     CONNECTED     78000    1/systemd            /run/systemd/journal/stdout
unix  3      [ ]         STREAM     CONNECTED     67808    3568/dbus-daemon     /run/user/42/bus
unix  3      [ ]         STREAM     CONNECTED     71805    2460/dbus-daemon     /run/dbus/system_bus_socket
unix  3      [ ]         STREAM     CONNECTED     55801    4058/gsd-keyboard
unix  3      [ ]         STREAM     CONNECTED     78014    1/systemd            /run/systemd/journal/stdout
unix  3      [ ]         STREAM     CONNECTED     17080    6837/qemu-kvm
unix  3      [ ]         STREAM     CONNECTED     78018    1/systemd            /run/systemd/journal/stdout
```

[root@vduhost ai-deploy-cluster-remoteworker]# journalctl -xe
```
Mar 25 22:29:40 vduhost.rtp.raleigh.ibm.com platform-python[28793]: ansible-command Invoked with _raw_params=aicli list cluster | grep >
Mar 25 22:29:40 vduhost.rtp.raleigh.ibm.com platform-python[28821]: ansible-command Invoked with _raw_params= _uses_shell=True warn=Tru>
Mar 25 22:29:41 vduhost.rtp.raleigh.ibm.com platform-python[28852]: ansible-command Invoked with _raw_params=aicli download iso vdu-day>
Mar 25 22:29:42 vduhost.rtp.raleigh.ibm.com sudo[28582]: pam_unix(sudo:session): session closed for user root
Mar 25 22:30:01 vduhost.rtp.raleigh.ibm.com sudo[28937]:     root : TTY=pts/4 ; PWD=/root/ztp-cluster-deploy/ai-deploy-cluster-remotewo>
Mar 25 22:30:01 vduhost.rtp.raleigh.ibm.com sudo[28937]: pam_systemd(sudo:session): Cannot create session: Already running in a session>
Mar 25 22:30:01 vduhost.rtp.raleigh.ibm.com sudo[28937]: pam_unix(sudo:session): session opened for user root by (uid=0)
Mar 25 22:30:02 vduhost.rtp.raleigh.ibm.com platform-python[28963]: ansible-setup Invoked with gather_subset=['all'] gather_timeout=10 >
Mar 25 22:30:04 vduhost.rtp.raleigh.ibm.com platform-python[29150]: ansible-stat Invoked with path=/root/racadm.sh follow=False get_che>
Mar 25 22:30:04 vduhost.rtp.raleigh.ibm.com platform-python[29157]: ansible-copy Invoked with src=/root/.ansible/tmp/ansible-tmp-161672>
Mar 25 22:30:04 vduhost.rtp.raleigh.ibm.com platform-python[29176]: ansible-command Invoked with _raw_params= _uses_shell=True warn=Tru>
Mar 25 22:30:12 vduhost.rtp.raleigh.ibm.com sudo[28937]: pam_unix(sudo:session): session closed for user root
Mar 25 22:38:18 vduhost.rtp.raleigh.ibm.com systemd[1]: Starting The Apache HTTP Server...
-- Subject: Unit httpd.service has begun start-up
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
--
-- Unit httpd.service has begun starting up.
Mar 25 22:38:18 vduhost.rtp.raleigh.ibm.com httpd[31086]: (98)Address already in use: AH00072: make_sock: could not bind to address [::>
Mar 25 22:38:18 vduhost.rtp.raleigh.ibm.com httpd[31086]: (98)Address already in use: AH00072: make_sock: could not bind to address 0.0>
Mar 25 22:38:18 vduhost.rtp.raleigh.ibm.com httpd[31086]: no listening sockets available, shutting down
Mar 25 22:38:18 vduhost.rtp.raleigh.ibm.com httpd[31086]: AH00015: Unable to open logs
Mar 25 22:38:18 vduhost.rtp.raleigh.ibm.com systemd[1]: httpd.service: Main process exited, code=exited, status=1/FAILURE
Mar 25 22:38:18 vduhost.rtp.raleigh.ibm.com systemd[1]: httpd.service: Failed with result 'exit-code'.
-- Subject: Unit failed
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
--
-- The unit httpd.service has entered the 'failed' state with result 'exit-code'.
Mar 25 22:38:18 vduhost.rtp.raleigh.ibm.com systemd[1]: Failed to start The Apache HTTP Server.
-- Subject: Unit httpd.service has failed
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
--
-- Unit httpd.service has failed.
--
-- The result is failed.
```

Then to unblock on worker node addition part, I killed haproxy service and started httpd
```
[root@vcuhost ai-deploy-cluster-remoteworker]# ps -ef | grep haproxy
root       26720       1  0 22:29 ?        00:00:00 /usr/bin/conmon --api-version 1 -c 0fa81684c2c04cd67e655faccd557a46c401bc908445c77da602e02941d7d40e -u 0fa81684c2c04cd67e655faccd557a46c401bc908445c77da602e02941d7d40e -r /usr/bin/runc -b /var/lib/containers/storage/overlay-containers/0fa81684c2c04cd67e655faccd557a46c401bc908445c77da602e02941d7d40e/userdata -p /var/run/containers/storage/overlay-containers/0fa81684c2c04cd67e655faccd557a46c401bc908445c77da602e02941d7d40e/userdata/pidfile -n haproxy_vcu --exit-dir /var/run/libpod/exits --socket-dir-path /var/run/libpod/socket -s -l k8s-file:/var/lib/containers/storage/overlay-containers/0fa81684c2c04cd67e655faccd557a46c401bc908445c77da602e02941d7d40e/userdata/ctr.log --log-level error --runtime-arg --log-format=json --runtime-arg --log --runtime-arg=/var/run/containers/storage/overlay-containers/0fa81684c2c04cd67e655faccd557a46c401bc908445c77da602e02941d7d40e/userdata/oci-log --conmon-pidfile /var/run/containers/storage/overlay-containers/0fa81684c2c04cd67e655faccd557a46c401bc908445c77da602e02941d7d40e/userdata/conmon.pid --exit-command /usr/bin/podman --exit-command-arg --root --exit-command-arg /var/lib/containers/storage --exit-command-arg --runroot --exit-command-arg /var/run/containers/storage --exit-command-arg --log-level --exit-command-arg error --exit-command-arg --cgroup-manager --exit-command-arg systemd --exit-command-arg --tmpdir --exit-command-arg /var/run/libpod --exit-command-arg --runtime --exit-command-arg runc --exit-command-arg --storage-driver --exit-command-arg overlay --exit-command-arg --storage-opt --exit-command-arg overlay.mountopt=nodev,metacopy=on --exit-command-arg --events-backend --exit-command-arg file --exit-command-arg container --exit-command-arg cleanup --exit-command-arg 0fa81684c2c04cd67e655faccd557a46c401bc908445c77da602e02941d7d40e
root       26732   26720  0 22:29 ?        00:00:00 /usr/local/sbin/haproxy-systemd-wrapper -p /run/haproxy.pid -db -f /usr/local/etc/haproxy/haproxy.cfg
root       26749   26732  0 22:29 ?        00:00:00 /usr/local/sbin/haproxy-master
root       26751   26749  0 22:29 ?        00:00:00 /usr/local/sbin/haproxy -p /run/haproxy.pid -db -f /usr/local/etc/haproxy/haproxy.cfg -Ds
root       31454    4261  0 22:43 pts/3    00:00:00 grep --color=auto haproxy
[root@vcuhost ai-deploy-cluster-remoteworker]#
[root@vcuhost ai-deploy-cluster-remoteworker]#
[root@vcuhost ai-deploy-cluster-remoteworker]#
[root@vcuhost ai-deploy-cluster-remoteworker]# kill -9 26751
[root@vcuhost ai-deploy-cluster-remoteworker]#
```

httpd service came up:
```
[root@vcuhost ai-deploy-cluster-remoteworker]#  service httpd restart
Redirecting to /bin/systemctl restart httpd.service
[root@vcuhost ai-deploy-cluster-remoteworker]#
[root@vcuhost ai-deploy-cluster-remoteworker]#
[root@vcuhost ai-deploy-cluster-remoteworker]#  service httpd status
Redirecting to /bin/systemctl status httpd.service
● httpd.service - The Apache HTTP Server
   Loaded: loaded (/usr/lib/systemd/system/httpd.service; disabled; vendor preset: disabled)
   Active: active (running) since Thu 2021-03-25 22:44:27 EDT; 6s ago
     Docs: man:httpd.service(8)
 Main PID: 31640 (httpd)
   Status: "Started, listening on: port 80"
    Tasks: 213 (limit: 2465614)
   Memory: 46.0M
   CGroup: /system.slice/httpd.service
           ├─31640 /usr/sbin/httpd -DFOREGROUND
           ├─31641 /usr/sbin/httpd -DFOREGROUND
           ├─31642 /usr/sbin/httpd -DFOREGROUND
           ├─31643 /usr/sbin/httpd -DFOREGROUND
           └─31644 /usr/sbin/httpd -DFOREGROUND

Mar 25 22:44:27 vcuhost.rtp.raleigh.ibm.com systemd[1]: Starting The Apache HTTP Server...
Mar 25 22:44:27 vcuhost.rtp.raleigh.ibm.com systemd[1]: Started The Apache HTTP Server.
Mar 25 22:44:27 vcuhost.rtp.raleigh.ibm.com httpd[31640]: Server configured, listening on: port 80
```

**Comment by yrobla** [comment]
I know this is an issue, and I think that i will just remove the expose-endpoints tag. Can you deploy successfully without the expose-endpoints?

**Comment by maheshd2** [comment]
@yrobla Yes, if we remove/skip expose endpoints role, this issue can be avoided.

**Comment by yrobla** [comment]
ok i'll check on how can we avoid that, i think best option should be changing the default port for haproxy, as it's the optional service. I'll update on it

**Comment by pmallan** [comment]
@yrobla any update on this?
